File: src/server_auto/routes_files.py
"""File operation endpoints for Server B."""

import os
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional, List
from ..config import BASE_DIR

router = APIRouter(prefix="/files")

# Allowed root for file operations
WORKSPACE = BASE_DIR


# --------------------------------------------------------------
class FileReadRequest(BaseModel):
    """Request to read a file."""
    path: str


# --------------------------------------------------------------
class FileWriteRequest(BaseModel):
    """Request to write a file."""
    path: str
    content: str


# --------------------------------------------------------------
class FileListRequest(BaseModel):
    """Request to list files."""
    path: Optional[str] = "."
    recursive: Optional[bool] = False


# --------------------------------------------------------------
def _safe_path(rel_path: str):
    """Resolve path safely within workspace."""
    resolved = (WORKSPACE / rel_path).resolve()
    if not str(resolved).startswith(
        str(WORKSPACE.resolve())
    ):
        return None
    return resolved


# --------------------------------------------------------------
@router.post("/read")
async def file_read(req: FileReadRequest):
    """Read a text file."""
    path = _safe_path(req.path)
    if path is None:
        return {
            "error": "path outside workspace"
        }
    if not path.exists():
        return {"error": f"not found: {req.path}"}
    if not path.is_file():
        return {"error": "not a file"}
    try:
        content = path.read_text(
            encoding="utf-8"
        )
        return {
            "path": req.path,
            "content": content,
            "chars": len(content),
        }
    except Exception as e:
        return {"error": str(e)}


# --------------------------------------------------------------
@router.post("/write")
async def file_write(req: FileWriteRequest):
    """Write or create a text file."""
    path = _safe_path(req.path)
    if path is None:
        return {
            "error": "path outside workspace"
        }
    try:
        path.parent.mkdir(
            parents=True, exist_ok=True
        )
        path.write_text(
            req.content, encoding="utf-8"
        )
        return {
            "status": "ok",
            "path": req.path,
            "chars": len(req.content),
        }
    except Exception as e:
        return {"error": str(e)}


# --------------------------------------------------------------
@router.get("/list")
async def file_list(
    path: str = ".",
    recursive: bool = False,
):
    """List files in a directory."""
    target = _safe_path(path)
    if target is None:
        return {
            "error": "path outside workspace"
        }
    if not target.exists():
        return {"error": f"not found: {path}"}
    if not target.is_dir():
        return {"error": "not a directory"}

    files = []
    if recursive:
        for f in sorted(target.rglob("*")):
            if f.is_file():
                rel = f.relative_to(WORKSPACE)
                files.append(str(rel))
    else:
        for f in sorted(target.iterdir()):
            rel = f.relative_to(WORKSPACE)
            suffix = "/" if f.is_dir() else ""
            files.append(str(rel) + suffix)

    return {"path": path, "files": files}

File: server_auto/src/routes_files.rs
/// File operation endpoints — /files/list, /files/read, /files/write
use axum::{routing::{get, post}, Router, Json};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::PathBuf;
use common::config;

pub fn router() -> Router {
    Router::new()
        .route("/files/list", get(file_list))
        .route("/files/read", post(file_read))
        .route("/files/write", post(file_write))
}

fn safe_path(rel: &str) -> Option<PathBuf> {
    let base = &config::get().base_dir;
    let p = base.join(rel);
    let resolved = p.canonicalize().unwrap_or_else(|_| p.clone());
    if resolved.starts_with(base) { Some(resolved) } else { None }
}

#[derive(Deserialize)]
struct ReadReq { path: String }

#[derive(Deserialize)]
struct WriteReq { path: String, content: String }

async fn file_list(
    axum::extract::Query(params): axum::extract::Query<std::collections::HashMap<String, String>>,
) -> Json<Value> {
    let rel = params.get("path").map(|s| s.as_str()).unwrap_or(".");
    let recursive = params.get("recursive").map(|s| s == "true").unwrap_or(false);
    let Some(target) = safe_path(rel) else {
        return Json(serde_json::json!({"error": "path outside workspace"}));
    };
    if !target.exists() {
        return Json(serde_json::json!({"error": format!("not found: {rel}")}));
    }
    if !target.is_dir() {
        return Json(serde_json::json!({"error": "not a directory"}));
    }
    let base = &config::get().base_dir;
    let mut files = Vec::new();
    if recursive {
        collect_recursive(&target, base, &mut files);
    } else {
        if let Ok(rd) = std::fs::read_dir(&target) {
            let mut entries: Vec<_> = rd.filter_map(|e| e.ok()).collect();
            entries.sort_by_key(|e| e.file_name());
            for entry in entries {
                let p = entry.path();
                let rel_p = p.strip_prefix(base).unwrap_or(&p);
                let suffix = if p.is_dir() { "/" } else { "" };
                files.push(format!("{}{suffix}", rel_p.display()));
            }
        }
    }
    Json(serde_json::json!({"path": rel, "files": files}))
}

fn collect_recursive(dir: &std::path::Path, base: &std::path::Path, out: &mut Vec<String>) {
    if let Ok(rd) = std::fs::read_dir(dir) {
        let mut entries: Vec<_> = rd.filter_map(|e| e.ok()).collect();
        entries.sort_by_key(|e| e.file_name());
        for entry in entries {
            let p = entry.path();
            if p.is_dir() {
                collect_recursive(&p, base, out);
            } else {
                let rel = p.strip_prefix(base).unwrap_or(&p);
                out.push(rel.display().to_string());
            }
        }
    }
}

async fn file_read(Json(req): Json<ReadReq>) -> Json<Value> {
    let Some(p) = safe_path(&req.path) else {
        return Json(serde_json::json!({"error": "path outside workspace"}));
    };
    match std::fs::read_to_string(&p) {
        Ok(content) => Json(serde_json::json!({
            "path": req.path,
            "content": content,
            "chars": content.len(),
        })),
        Err(e) => Json(serde_json::json!({"error": e.to_string()})),
    }
}

async fn file_write(Json(req): Json<WriteReq>) -> Json<Value> {
    let Some(p) = safe_path(&req.path) else {
        return Json(serde_json::json!({"error": "path outside workspace"}));
    };
    if let Some(parent) = p.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    match std::fs::write(&p, &req.content) {
        Ok(_) => Json(serde_json::json!({
            "status": "ok",
            "path": req.path,
            "chars": req.content.len(),
        })),
        Err(e) => Json(serde_json::json!({"error": e.to_string()})),
    }
}
